decisions/0004-sync-routes-from-k8s-to-cc.md

```diff
@@ -0,0 +1,113 @@
+
+##### Sync from k8s to CCDB
+1. Problem: CCDB ✅ & Route CRD ❌
+   1. Solution: Raise the error to the user, abort the CCDB transaction.
+1. Problem: Route CRD ✅ & CCDB ❌
+   1. Solution: Raise the error to the user, but leave the Route CRD. The "bulk sync loop" will not consider the route "reconciled" until it is correct in CCDB.
+
+### Viable Long Term Implementations & Tradeoffs
+
+##### CCDB->k8s Sync in Ruby
+1. Mimics ProcessesSync for efficient access to CCDB
+1. Run it in an independent process from clock_global to avoid BOSH entanglement
+1. Unclear what scale limitations we might encounter if we regularly download
+   all k8s Routes.
+
+##### CCDB->k8s Sync in Go
+1. Probably a new, independent go program
+1. Can take advantage of go-client cache & reflector to avoid repeatedly
+   downloading all k8s state.
+   1. It's unclear, however, that kubebuilder could help
+   1. What does it mean to "reconcile" a change to a single Route CRD if CCDB is authoritative?
+1. Would need to page through CF API to build picture of CF state.
+
+##### Sync from k8s->CCDB with kubebuilder controller and reloads
+1. Probably a RoutesController akin to our existing kpack BuildsController
+1. Can take advantage of go-client cache & reflector to avoid repeatedly
+   downloading all k8s state.
+   1. Can take advantage of existing kubebuilder patterns
+   1. "Reconciling" a change to a single Route CRD means verifying with the CF
+      API that it has been loaded from k8s.
+1. Route Reconcile must `POST /v3/route/:guid/reload` endpoint on CF API
+   1. These must only succeed if the given resource can be loaded from k8s
+   1. Each "reconcile" call requires a successful call to this endpoint or else
+      it'll be reqeued.
+   1. Endpoint must short circuit if provided k8s resourceVersion matches one
+      stored in CCDB. This avoids endless loops and avoids unnecessary k8s
+      traffic.
+1. In normal operation, CF API must be careful not to commit changes to CCDB
+   that failed to be applied to k8s.
+   1. In practice, this means k8s Routes operations must occur inside CCDB
+      transactions.
+1. In failure modes, restarting RoutesController would resync existing Routes.
+1. Hard to handle Routes that are missing in k8s but still present in CCDB.
+   1. In theory this should never happen as the k8s delete would need to be
+      reconciled, but in practice it could happen due to bugs
+   1. Might require us to periodically `POST /v3/routes/:guid/reload` every route
+      in CCDB, anyways, or just have operators do that in the rare case the
+      route counts are out-of-sync.
+   1. Maybe PATCHes cause deletes with 409 Gone?
+1. Allows platform operators to add routes as part of their cf-for-k8s
+   configuration that could be viewable via the CF API.
+
+## Decision
+
+I'm inclined towards trying out k8s->CCDB sync, despite its newness. It goes a
+step further to embrace kubernetes and allows us to make heavy use of existing
```

Comment:
One question I have in general that I'm not sure I've seen discussed elsewhere is are we on board with going all in on Kubernetes such that it affects CCDB's schema?

Cloud Controller has already been replatformed from DEA -> Diego and now to Kubernetes...
I think by going down this road we're tying ourselves pretty strongly to Kubernetes and if there's ever a _next platform_ Cloud Controller might not be able to make it there without significant rework after this.

Just wondering what the thoughts are around this.

Reply:
Every one of those moves has required significant rework, even when CF Engineers were the folks designing the replacement. I'm not super convinced that changing the logical API-store-of-record is the big change that will inextricably couple the ecosystem to k8s. 

That said, treating CCDB as a reloadable cache ties us to k8s much less strongly than having the CF-API directly backed by the k8s API, and we're already moving forward with some feature sets (kpack buildpacks) that _are_ backed directly by a new schema in the k8s API. So long as the CF API can `GET /v3/apps` without chatting extensively to k8s, the API can probably still be modified to work with a new container runtime. If that stops being possible, I share your concern that it won't be long before the CF API is inextricably coupled to the data model of wherever we get our `/v3/apps` and `/v3/processes`.

Reply:
I'd argue that we'd couple capi (and other components) to the k8s api-server and the edge driven reconciliation loop as a means of communication between the CF components. Right now those are coupled to SQL-based data storage and multiple home-grown reconciliation loops/bulk sync agents.

There is no coupling to k8s as a container scheduler in using CRDs for data storage and state coordination. One could even think of a k8s-api BOSH release and moving the components over to that model for both cf4bosh and cf4k8s.